src/snowflake/snowpark/mock/_functions.py

```diff
@@ -863,59 +859,141 @@ def mock_to_char(
+            time_str = try_convert(
+                lambda x: datetime.date.strftime(x, format), try_cast, data
+            ).strip()
+            # python doesn't offer a way to control digits in microseconds
+            # when converting datatime into string using format %f, hence here we manually control the output fractional
+            # microsecond parts
+            # we find the beginning of 6 consecutive digits and manipulate the string
+            # CAVEAT: this solution can not handle cases yyyymmddhhmmss
+            if "%f" in format:
+                pattern = r"\d{6}"
+                if len(re.findall(pattern, time_str)) > 1:
+                    # if we detect multiple consecutive 6 digits, it's a format we can't handle
+                    LocalTestOOBTelemetryService.get_instance().log_not_supported_error(
+                        external_feature_name=f"Use format string {_fmt} with TimestampType in TO_CHAR",
+                        internal_feature_name="mock_to_char",
+                        parameters_info={
+                            "source_datatype": type(source_datatype).__name__,
+                            "fmt": str(_fmt),
+                        },
+                        raise_error=NotImplementedError,
+                    )
+                start_idx = re.search(r"\d{6}", time_str).start()
+                end_idx = start_idx + 6
+                # truncate the microsecond string
+                fractional_seconds_str = time_str[start_idx:end_idx][
+                    0 : min(6, fractional_seconds)
+                ]
+                # concatenate the whole string
+                time_str = (
+                    time_str[:start_idx] + fractional_seconds_str + time_str[end_idx:]
+                )
+            return time_str
+        elif isinstance(source_datatype, BinaryType):
+            _fmt = (_fmt or "HEX").upper()
+            fmt_decoder = {
+                "HEX": binascii.hexlify,
+                "BASE64": base64.b64encode,
+                "UTF-8": lambda x: x,
+            }.get(_fmt)
+
+            if fmt_decoder is None:
+                raise SnowparkSQLException(f"Invalid binary format {fmt}")
+            return try_convert(fmt_decoder, try_cast, data).decode()
+        elif isinstance(source_datatype, BooleanType):
+            return try_convert(lambda x: str(x).lower(), try_cast, data)
+        elif isinstance(source_datatype, (VariantType, ArrayType, MapType)):
+            from snowflake.snowpark.mock import CUSTOM_JSON_ENCODER
+
+            # here we reuse CUSTOM_JSON_ENCODER to dump a python object to string, by default json dumps added
+            # double quotes to the output which we do not need in output, we strip the beginning and ending double quote.
+            return try_convert(
+                lambda x: json.dumps(
+                    x, cls=CUSTOM_JSON_ENCODER, separators=(",", ":")
+                ).strip('"'),
```

Comment:
Doesn't json.dumps only add double quotes when the incoming value is a str, in that case, how do we differentiate when there is a str in a Variant column that is "[]", or when there is an array stored in the variant column? Or do we need to distinguish at all?

Reply:
are you asking about the `strip('"')` here? this is to remove the first and last double quotes around the output string -- e.g., '"xxxxxxxxxxxx"' -> 'xxxxxxxxxxxx', not to interfere the values or fields.

specially this is to handle cases in which we try to dump string: e.g., json.dumps("123") would return a quoted string '"123"', this has no side-effect to other types.

I will leave a notes here.

Reply:
based upon test, snowflake doesn't differentiate `select to_char(123::variant)` or `select to_char('123':variant)` -- both returns str `123`, without being quoted. so we are good here